pkg/providers/aws/provider_postgressnapshot.go

```diff
@@ -0,0 +1,244 @@
+			return nil, croType.StatusMessage(errMsg), errorUtil.New(errMsg)
+		}
+		p.logger.Info("creating rds snapshot")
+		_, err = rdsSvc.CreateDBSnapshot(&rds.CreateDBSnapshotInput{
+			DBInstanceIdentifier: aws.String(instanceName),
+			DBSnapshotIdentifier: aws.String(snapshotName),
+		})
+		if err != nil {
+			errMsg := "error creating rds snapshot"
+			return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+		}
+		return nil, "snapshot started", nil
+	}
+
+	// if snapshot status complete update status
+	if *foundSnapshot.Status == "available" {
+		return &providers.PostgresSnapshotInstance{
+			Name: *foundSnapshot.DBSnapshotIdentifier,
+		}, "snapshot created", nil
+	}
+
+	// creation in progress
+	msg := fmt.Sprintf("current snapshot status : %s", *foundSnapshot.Status)
+	p.logger.Info(msg)
+	return nil, croType.StatusMessage(msg), nil
+}
+
+func (p *PostgresSnapshotProvider) deletePostgresSnapshot(ctx context.Context, snapshot *v1alpha1.PostgresSnapshot, postgres *v1alpha1.Postgres, rdsSvc rdsiface.RDSAPI) (croType.StatusMessage, error) {
+	snapshotName := snapshot.Status.SnapshotID
+	foundSnapshot, err := p.findSnapshotInstance(rdsSvc, snapshotName)
+
+	if err != nil {
+		errMsg := "failed to describe snaphots in AWS"
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// snapshot is deleted
+	if foundSnapshot == nil {
+		resources.RemoveFinalizer(&snapshot.ObjectMeta, DefaultFinalizer)
+
+		if err := p.client.Update(ctx, snapshot); err != nil {
+			msg := "failed to update instance as part of finalizer reconcile"
+			return croType.StatusMessage(msg), errorUtil.Wrapf(err, msg)
+		}
+		return "snapshot deleted", nil
+	}
+
+	deleteSnapshotInput := &rds.DeleteDBSnapshotInput{
+		DBSnapshotIdentifier: foundSnapshot.DBSnapshotIdentifier,
+	}
+
+	deleteSnapshotOutput, err := rdsSvc.DeleteDBSnapshot(deleteSnapshotInput)
+
+	if err != nil {
+		errMsg := fmt.Sprintf("failed to delete snapshot %s in aws", snapshotName)
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	p.logger.Debugf("delete snapshot output %+v", deleteSnapshotOutput)
+
+	return "snapshot deletion started", nil
+}
+
+func (p *PostgresSnapshotProvider) findSnapshotInstance(rdsSvc rdsiface.RDSAPI, snapshotName string) (*rds.DBSnapshot, error) {
+	// check snapshot exists
+	listOutput, err := rdsSvc.DescribeDBSnapshots(&rds.DescribeDBSnapshotsInput{
+		DBSnapshotIdentifier: aws.String(snapshotName),
+	})
+	if err != nil {
+		rdsErr, isAwsErr := err.(awserr.Error)
+		if isAwsErr && rdsErr.Code() == "DBSnapshotNotFound" {
+			return nil, nil
+		}
+		return nil, err
+	}
+	var foundSnapshot *rds.DBSnapshot
+	for _, c := range listOutput.DBSnapshots {
+		if *c.DBSnapshotIdentifier == snapshotName {
+			foundSnapshot = c
+			break
+		}
+	}
+	return foundSnapshot, nil
+}
+
+func (r *PostgresSnapshotProvider) getAwsSession(ctx context.Context, postgresCr *v1alpha1.Postgres) (*session.Session, error) {
```

Comment:
This functionality used to be directly in the create function but after adding the ability to delete, it made sense to make it a bit more reusable. The thing is I also have this function in `provider_redissnapshot.go` and it's pretty much identical except it takes in a Redis CR type as an argument. Question: are we happy with the duplication or should I make it into a more generic function that will work for both providers?